daemon/api_systems_test.go

```diff
@@ -1228,3 +1232,648 @@ func (s *systemsSuite) TestSystemInstallActionError(c *check.C) {
+			"snap-size":     strconv.Itoa(int(size)),
+			"snap-revision": "10",
+		})
+
+		// this is required right now. should it be?
+		decl := s.mockStoreAssertion(c, asserts.SnapDeclarationType, map[string]interface{}{
+			"series":       "16",
+			"snap-id":      snaptest.AssertedSnapID(name),
+			"snap-name":    name,
+			"publisher-id": s.dev1acct.AccountID(),
+			"timestamp":    time.Now().Format(time.RFC3339),
+		})
+
+		assertions = append(assertions, string(asserts.Encode(rev)), string(asserts.Encode(decl)))
+
+		content, err := os.ReadFile(f)
+		c.Assert(err, check.IsNil)
+
+		snapFormData[name] = string(content)
+	}
+
+	valSetString := accountID + "/" + validationSet
+	fields := map[string][]string{
+		"action":          {"create"},
+		"assertion":       assertions,
+		"label":           {expectedLabel},
+		"validation-sets": {valSetString},
+	}
+
+	form, boundary := createFormData(c, fields, snapFormData)
+
+	daemon.MockDevicestateCreateRecoverySystem(func(st *state.State, label string, opts devicestate.CreateRecoverySystemOptions) (*state.Change, error) {
+		c.Check(expectedLabel, check.Equals, label)
+		c.Check(opts.ValidationSets, check.HasLen, 1)
+		c.Check(opts.ValidationSets[0].Body(), check.DeepEquals, vsetAssert.Body())
+
+		c.Check(opts.LocalSnaps, check.HasLen, 3)
+
+		for _, vs := range opts.ValidationSets {
+			c.Check(vs.AccountID(), check.Equals, accountID)
+		}
+
+		return st.NewChange("change", "..."), nil
+	})
+
+	req, err := http.NewRequest("POST", "/v2/systems", &form)
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/form-data; boundary="+boundary)
+	req.Header.Set("Content-Length", strconv.Itoa(form.Len()))
+
+	res := s.asyncReq(c, req, nil)
+
+	st := s.d.Overlord().State()
+	st.Lock()
+	defer st.Unlock()
+
+	c.Check(st.Change(res.Change), check.NotNil)
+}
+
+func (s *systemsCreateSuite) TestCreateSystemActionOfflinePreinstalled(c *check.C) {
+	const (
+		expectedLabel = "1234"
+	)
+
+	fields := map[string][]string{
+		"action": {"create"},
+		"label":  {expectedLabel},
+	}
+
+	form, boundary := createFormData(c, fields, nil)
+
+	daemon.MockDevicestateCreateRecoverySystem(func(st *state.State, label string, opts devicestate.CreateRecoverySystemOptions) (*state.Change, error) {
+		c.Check(expectedLabel, check.Equals, label)
+		c.Check(opts.ValidationSets, check.HasLen, 0)
+		c.Check(opts.LocalSnaps, check.HasLen, 0)
+		c.Check(opts.Offline, check.Equals, true)
+
+		return st.NewChange("change", "..."), nil
+	})
+
+	req, err := http.NewRequest("POST", "/v2/systems", &form)
+	c.Assert(err, check.IsNil)
+	req.Header.Set("Content-Type", "multipart/form-data; boundary="+boundary)
+	req.Header.Set("Content-Length", strconv.Itoa(form.Len()))
+
+	res := s.asyncReq(c, req, nil)
+
+	st := s.d.Overlord().State()
+	st.Lock()
+	defer st.Unlock()
+
+	c.Check(st.Change(res.Change), check.NotNil)
+}
+
+func (s *systemsCreateSuite) TestCreateSystemActionOfflineJustValidationSets(c *check.C) {
```

Comment:
from this POV I could see how it could make sense to support passing a "offline": "true" flag via JSON as well

Reply:
Yes, I agree. Do you want that in this PR?

Reply:
It would also make it more consistent with the offline remodel variant

Reply:
if it doesn't add too many lines? otherwise a follow-up is also fine

Reply:
Made this change in aedcb639bba0dadd40eca62bfdf421370a79ca88.